presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java

```diff
@@ -143,34 +147,45 @@ public StructSelectiveStreamReader(
                 .map(Subfield.NestedField::getName)
                 .collect(toImmutableSet());
 
-        if (!checkMissingFieldFilters(nestedStreams, filters)) {
+        if (!checkMissingFieldFilters(nestedStreams.values(), filters)) {
             this.missingFieldFilterIsFalse = true;
             this.nestedReaders = ImmutableMap.of();
             this.orderedNestedReaders = new SelectiveStreamReader[0];
         }
         else if (outputRequired || !fieldsWithFilters.isEmpty()) {
             ImmutableMap.Builder<String, SelectiveStreamReader> nestedReaders = ImmutableMap.builder();
-            for (int i = 0; i < nestedStreams.size(); i++) {
-                StreamDescriptor nestedStream = nestedStreams.get(i);
-                String fieldName = nestedStream.getFieldName().toLowerCase(Locale.ENGLISH);
-                Optional<Type> fieldOutputType = nestedTypes.isPresent() ? Optional.of(nestedTypes.get().get(i)) : Optional.empty();
+            Map<String, Field> nestedTypes = outputType.isPresent() ? ((RowType) this.outputType).getFields().stream()
+                    .collect(toImmutableMap((f) -> f.getName()
+                            .orElseThrow(() -> new IllegalArgumentException(
+                                    "ROW type does not have field names declared: " + this.outputType))
+                            .toLowerCase(Locale.ENGLISH), Function.identity()))
+                    : ImmutableMap.of();
+            Set<String> structFields = outputType.isPresent() ? nestedTypes.keySet() : nestedStreams.keySet();
+            for (String fieldName : structFields) {
+                StreamDescriptor nestedStream = nestedStreams.get(fieldName);
                 boolean requiredField = requiredFields.map(names -> names.containsKey(fieldName)).orElse(outputRequired);
+                Optional<Type> fieldOutputType = nestedTypes.get(fieldName) != null ? Optional.of(nestedTypes.get(fieldName).getType()) : Optional.empty();
```

Comment:
use Optional.ofNullable

Reply:
the Optional.empty is based on nestedTypes containing the fieldName, if its present then I m retrieving the subfield type. nestesTypes.get(field).geType

Reply:
Optional.ofNullable(nestedTypes.get(fieldName)).map(Field::getType)